service/client_state.cc

```diff
@@ -224,38 +224,77 @@ future<> service::client_state::has_access(const sstring& ks, auth::command_desc
                 ks + " can be granted only SELECT or DESCRIBE permissions to a non-superuser.");
     }
 
+    if (cmd.resource.kind() == auth::resource_kind::data && cmd.permission == auth::permission::SELECT && cmd.resource.is_vector_indexed_table()) {
+
+        co_return co_await ensure_has_permission<auth::command_desc_with_permission_set>({auth::permission_set::of<auth::permission::SELECT, auth::permission::VECTOR_SEARCH_INDEXING>(), cmd.resource});
+
+    }
+
     co_return co_await ensure_has_permission(cmd);
 }
 
-future<bool> service::client_state::check_has_permission(auth::command_desc cmd) const {
+bool intersects_permissions(const auth::permission_set& permissions, const auth::command_desc_with_permission_set& cmd) {
+    return permissions.intersects(cmd.permission);
+}
+
+bool intersects_permissions(const auth::permission_set& permissions, const auth::command_desc& cmd) {
+    return permissions.contains(cmd.permission);
+}
+
+sstring service::client_state::generate_authorization_error_msg(const auth::command_desc& cmd) const {
+    return format("User {} has no {} permission on {} or any of its parents",
+            *_user,
+            auth::permissions::to_string(cmd.permission),
+            cmd.resource);
+}
+
+sstring service::client_state::generate_authorization_error_msg(const auth::command_desc_with_permission_set& cmd) const {
+    std::ostringstream os;
+    for (auto p : auth::permissions::to_strings(cmd.permission)) {
+        if (os.tellp() > 0) {
+            os << ", ";
+        }
+        os << p;
+    }
+    sstring perm_names = os.str();
+    return format("User {} has none of the permissions ({}) on {} or any of its parents",
+            *_user,
+            perm_names,
+            cmd.resource);
+}
+
+template <typename Cmd>
+future<bool> service::client_state::check_has_permission(Cmd cmd) const {
     if (_is_internal) {
         co_return true;
     }
 
     std::optional<auth::resource> parent_r = cmd.resource.parent();
 
     auth::permission_set set = co_await auth::get_permissions(*_auth_service, *_user, cmd.resource);
-    if (set.contains(cmd.permission)) {
+    if (intersects_permissions(set, cmd)) {
         co_return true;
     }
     if (parent_r) {
-        co_return co_await check_has_permission({cmd.permission, *parent_r});
+        co_return co_await check_has_permission<Cmd>({cmd.permission, *parent_r});
     }
     co_return false;
 }
+template future<bool> service::client_state::check_has_permission<auth::command_desc>(auth::command_desc) const;
+template future<bool> service::client_state::check_has_permission<auth::command_desc_with_permission_set>(auth::command_desc_with_permission_set) const;
 
-future<> service::client_state::ensure_has_permission(auth::command_desc cmd) const {
+template <typename Cmd>
```

Comment:
maybe worth put a constraint that it's either of those two cmd types we have? not a must have though